documentation/architecture/messaging/Reliability.md

```diff
@@ -0,0 +1,201 @@
+# Reliability and uniqness
+
+Reliability describe how many sent messages were received and whether some messages are lost
+
+Reliability can be described as a fraction of a number of unique messages received `Ru` to a number of messages sent `S`
+
+When sending messages on `A->B`, when `A` sends `SA` messages and `B` receives `RuB` unique messages, reliability of such delivery is `R(A->B) = RuB/SA =< 1`
+
+Given that `Ru` cannot be higher than `S`, reliability is always `=< 1`
+
+Uniqness describes whether messages received correspond to unique sent messages or are duplicates of the same sent message
+
+We can measure duplication rate in a delivery by calculating a freaction of total messages received `R` to unique messages received `Ru`. `DR = R/Ru`
+Duplication rate of `1` means each received message was received only once, rate of `2` means that messages were received 2 times on average, not necessary each message was duplicated though.
+
+For consistency wirh reliability we can use inverted duplication rate, call it **uniqness rate** as `UR = 1/DR = Ru/R`
+
+If `UR = 1` when each message was received only one time, `UR = 0.5` when messages are received twice on average
+
+<img src="./images/reliability.jpg" width="100%">
+
+Reliability `R(A->B) = 3/4` - 1 message out of 4 was lost
+Uniqness rate `UR(A->B) = 3/4` - 1 message out of 4 was duplicated
+
+
+## Delivery modes
+
+Reliability often described together with uniqness as "delivery modes"
+
+If we have a delivery mode as a tuple of reliability and uniqness rate: `DM(A->B) = {R(A->B), UR(A->B)}`
+then we can describe the following modes:
+
+1. Unreliable. When messages may be lost or duplicated `DM = {R < 1, UR < 1}`
+1. At most once. When all messages are unique, but may be missing `DM = {R < 1, UR = 1}`
+1. At least once. When all messages are received, but may have duplicates `DM = {R = 1, UR < 1}`
+1. Exactly once. When all messages are received and only once `DM = {R = 1, UR = 1}`
+
+Exactly once delivery is controversal and hard to achieve in presence of errors.
+For practical purposes we're going to use more relaxed version of this definition
+where `DM = {R -> 1, UR -> 1}`
+
+<img src="./images/delivery_modes.jpg" width="100%">
```

Comment:
There's a repeated typo in this image on `receivzed`

Reply:
Can't find it. Must have been addressed already.